feat: initial setup

cdunster opened the pull request.

cdunster committed:
feat: add pulumi config to create droplet and install nomad

cdunster committed:
chore: add TODO for DigitalOcean token

ThetaSinner commented on cloud-init.yaml:
```diff
@@ -0,0 +1,6 @@
+#cloud-config
+runcmd:
```
Ooh cloudinit, nice

cdunster replied:
Do you know it well? Because I do not so there may be many mistakes :smile:

ThetaSinner replied:
Not very well, I've used it a few times from the manual and I think it's a great tool. Definitely wouldn't claim to know it :D

cdunster replied:
Fair enough, then I guess we see how it goes :smile:

ThetaSinner commented on go.mod:
```diff
@@ -0,0 +1,93 @@
+module nomad-server
+
+go 1.21
```
Any chance of using a newer Go version? 1.23.x ideally

ThetaSinner commented on main.go:
```diff
@@ -0,0 +1,32 @@
+func main() {
+	userData, err := os.ReadFile("cloud-init.yaml")
+	if err != nil {
+		return
```
`log.Fatalf`? Or just `fmt.Println` the error

cdunster committed:
chore: install nomad via cloud-init packages

cdunster committed:
chore: remove sudo when enablinga and starting nomad

cdunster committed:
docs: add development section to the README

cdunster committed:
chore: add CI workflow to preview Pulumi changes on PR to main

holochain-release-automation2 commented:
#### :tropical_drink: `preview` on nomad-server/holochain/nomad-server

<details>
<summary>Pulumi report</summary>


<pre>
Previewing update (holochain/nomad-server)

View Live: https://app.pulumi.com/holochain/nomad-server/nomad-server/previews/0ef27af9-96bc-4ac8-bd37-cb9cc74354ad

+ pulumi:pulumi:Stack: (create)
[urn=urn:pulumi:nomad-server::nomad-server::pulumi:pulumi:Stack::nomad-server-nomad-server]
+ digitalocean:index/droplet:Droplet: (create)
    [urn=urn:pulumi:nomad-server::nomad-server::digitalocean:index/droplet:Droplet::nomad-server-01]
    [provider=urn:pulumi:nomad-server::nomad-server::pulumi:providers:digitalocean::default_4_38_0::04da6b54-80e4-46f7-96ec-b56ff0331ba9]
    backups         : false
    gracefulShutdown: false
    image           : "ubuntu-24-04-x64"
    ipv6            : true
    monitoring      : false
    name            : "nomad-server-01"
    region          : "ams3"
    resizeDisk      : true
    size            : "s-1vcpu-512mb-10gb"
    sshKeys         : [
        [0]: "41:9b:5f:a0:f5:18:d1:ae:0f:13:04:e0:17:7b:a8:58"
        [1]: "9d:50:aa:17:2a:ba:63:c5:58:6f:35:a2:fe:89:cf:ec"
        [2]: "97:5e:c1:c1:c8:63:3a:b2:34:62:40:28:df:25:99:63"
        [3]: "7f:23:ae:c8:23:58:36:35:93:02:8f:5d:1c:ff:86:2a"
        [4]: "ff:0e:80:0a:47:aa:34:c7:ef:71:d2:63:70:2b:c1:72"
        [5]: "36:ca:c9:95:96:ee:d5:e1:d2:ff:52:6a:32:04:78:2b"
        [6]: "38:ad:e4:d0:64:e0:3b:06:c0:17:5e:95:c9:4f:22:db"
        [7]: "7d:5e:86:00:fc:f0:9c:86:56:6f:85:ef:71:de:23:09"
    ]
    tags            : [
        [0]: "nomad"
    ]
    userData        : (yaml) {
        packages   : [
            [0]: "nomad"
        ]
        runcmd     : [
            [0]: "nomad agent -c /etc/nomad/server.conf.hcl"
        ]
        write_files: [
            [0]: {
                content: "data_dir  = \"/var/lib/nomad\"

server {
  enabled          = true
  bootstrap_expect = 1 # should increase this after testing
}
"
                path   : "/etc/nomad/server.conf.hcl"
            }
        ]
    }

Resources:
+ 2 to create
</pre>

</details>

cdunster committed:
chore: update to latest go with bundles a toolchain

cdunster committed:
feat: log error and exit if fail to read user data file

cdunster committed:
feat: update resource name to have an ID of 01

We will likely need more droplets running Nomad server in the future so
these will need to be numbered.

cdunster committed:
feat: give all SSH keys on DigitalOcean access to the droplet

cdunster committed:
feat: add DigitalOcean personal access token

cdunster committed:
docs: add section about updating DigitalOcean PAT in README

cdunster committed:
chore: add CI workflow to deploy the changes to Pulumi

cdunster committed:
docs: add section in README about using PRs to apply changes

cdunster committed:
chore: add CI job to check the cloud-init file for lints

cdunster committed:
feat: create Nomad server config with cloud-init

cdunster committed:
chore: replace cloud-init-lint CI action with bash script

The action is broken and unmaintained.

cdunster committed:
chore: mark cloud-init lint job as required to pass CI

cdunster committed:
chore: add result of cloud-init lint job to summary

cdunster committed:
chore: add errors as messages in cloud-init lint job

cdunster commented on .github/workflows/preview.yaml:
```diff
@@ -0,0 +1,56 @@
+          if ! cloud-init schema -c cloud-init.yaml 2> >(tee stderr.log) >> $GITHUB_STEP_SUMMARY
+          then
+            # Print errors as such in GitHub logs.
+            cat stderr.log | grep "in.*line.*column" -B1 -A1 --no-group-separator | sed -z 's/\([^\n]*\)\n\s*in "\([^"]*\)", line \([[:digit:]]\+\), column \([[:digit:]]\+\):\s*/::error file=cloud-init.yaml,title=\1 in \2,line=\3,col=\4::/g'
```
This leaves annotations on the PR if the lint fails but I can imagine that people might not want to understand and maintain my lovely sed string so I can remove if preferred.

ThetaSinner replied:
Haha, fancy :D Useful for now, if it breaks I will see how hard it is to fix

ThetaSinner commented on .github/workflows/preview.yaml:
```diff
@@ -0,0 +1,56 @@
+on:
+  pull_request:
+    branches:
+      - main
```
Generally prefer skipping the `branches` restriction but I guess in this repo we'll always be merging into `main`?

ThetaSinner commented on .github/workflows/preview.yaml:
```diff
@@ -0,0 +1,56 @@
+          stack-name: holochain/nomad-server
+          comment-on-pr: true
+          diff: true
+          github-token: ${{ secrets.HRA2_GITHUB_TOKEN }}
```
Are these provisioned with Pulumi? They can be if you want, don't have to be yet as I'm still figuring that out a bit

cdunster replied:
No... good catch! I immediately forgot about using Pulumi for this stuff

cdunster replied:
See: https://github.com/holochain/hc-github-config/pull/14

cdunster replied:
They are now managed by Pulumi :slightly_smiling_face:

ThetaSinner commented on main.go:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	pulumi.Run(func(ctx *pulumi.Context) error {
+		getSshKeysResult, err := digitalocean.GetSshKeys(ctx, &digitalocean.GetSshKeysArgs{}, nil)
```
Temporary thing, if we're going to give everyone access then we should finish questioning who is in that list... because there are some people in there who I think should be removed due to no longer working here. Currently none of their keys are linked to machines but this would do that I think

cdunster replied:
I looked at the list and thought it seemed reasonable, but maybe I looked in the wrong place.

ThetaSinner reviewed (dismissed):
Looking great. I have a few questions/comments but happy to approve at this point :)

jost-s commented on README.md:
````diff
@@ -1,3 +1,58 @@
+pulumi org set-default holochain
+```
+
+Finally, select the Pulumi stack that you want to use, for this repo it is `nomad-server`.
````
```suggestion
Finally, select the Pulumi stack that you want to use. For this repo it is `nomad-server`.
```

jost-s commented on README.md:
```diff
@@ -1,3 +1,58 @@
+is reviewed and merged into the `main` branch, a new workflow will push the
+changes to Pulumi.
+
+## Chaning the DigitalOcean token
```
```suggestion
## Changing the DigitalOcean token
```

jost-s commented on README.md:
```diff
@@ -1,3 +1,58 @@
+
+Pulumi requires a Personal Access Token (PAT) for DigitalOcean to make calls to the API.
+
+Currently the PAT is linked to the `cdunster` DigitalOcean user account. To
+change the token, run the following command:
```
```suggestion
To set the PAT of your user account, run the following command:
```

cdunster replied:
This is the DigitalOcean token that Pulumi will always use, the one that we share and is used in CI etc. I feel like your suggestion makes it seem like each person has their own PAT configured. At some point, we should probably give one of the `holochain_release_automation` bot accounts access to DO and change this token to that.

jost-s replied:
Okay. I get you now. So we're all going to use your PAT for the time being. Then it might not be necessary to mention how to change the token. But fine, just leave it as is.

cdunster replied:
Yeah, using my account isn't ideal. I would like to use the `holochain-release-automation2` account that we use elsewhere but I don't know if that has access to DigitalOcean as I can't even view other team members let alone add them.

ThetaSinner replied:
Follow up ticket for this? We can discuss with @zippy getting access for HRA2

cdunster replied:
Done, see: #2

cdunster committed:
chore: change preview CI workflow to run on all PRs

cdunster committed:
docs: fix typos and punctuation in README

Co-authored-by: Jost <28270981+jost-s@users.noreply.github.com>

c12i reviewed (commented):
This is more of question than a review and its regarding the naming of this repo, is this repo also going to house the nomad job files or does this only contain the pulumi configs. If its the latter, perhaps we could rename the repo to something more fitting like `nomad-server-config`

cdunster commented:
> This is more of question than a review and its regarding the naming of this repo, is this repo also going to house the nomad job files or does this only contain the pulumi configs. If its the latter, perhaps we could rename the repo to something more fitting like `nomad-server-config`

Good question, but unfortunately, I don't know the answer yet :confused:. I think that some jobs will be specified in this repo for things other than running Wind Tunnel scenarios, like aggregating results or something. But I really don't know how this repo will grow yet, it might just be the Pulumi config to deploy a Nomad server or it might manage all the Nomad servers used by Wind Tunnel or maybe it will manage all of our DigitalOcean droplets and become `droplet-config` or something.

cdunster commented on cloud-init.yaml:
```diff
@@ -0,0 +1,14 @@
+        bootstrap_expect = 1 # should increase this after testing
+      }
+runcmd:
+  - nomad agent -c /etc/nomad/server.conf.hcl
```
I hope that this runs every time the droplet boots but I guess we'll find out...
